src/datasets/arrow_dataset.py

```diff
@@ -2818,6 +2759,52 @@ def _map_single(
         # If set to False, no new arrow table will be created
         update_data = None
 
+        lazy_formatting = not input_columns and supports_lazy_formatting(self._format_type)
+        input_formatter = get_formatter(
+            self._format_type, features=self.features, lazy=lazy_formatting, **self._format_kwargs
+        )
+        if lazy_formatting:
+            # Save undecoded data (to preserve exact image/audio format for instance) after processing unless decoded data is modified in the `function`
+            lazy_type_to_extend = input_formatter.lazy_row_type if not batched else input_formatter.lazy_batch_type
+
+            class DictWithDecodableDataToRestore(lazy_type_to_extend):
+                def __init__(self, *args, **kwargs):
+                    super().__init__(*args, **kwargs)
+                    self.decodable_data_to_restore = {}
+
+                def __getitem__(self, key):
+                    if (
+                        key in self.cols_to_decode
+                        and self.formatter.features._column_requires_decoding[key]
+                        and key not in self.decodable_data_to_restore
+                    ):
+                        self.decodable_data_to_restore[key] = self.data[key]
+                    return super().__getitem__(key)
+
+                def __setitem__(self, key, value):
+                    if key in self.decodable_data_to_restore:
+                        del self.decodable_data_to_restore[key]
+                    super().__setitem__(key, value)
+
+                def __delitem__(self, key):
+                    if key in self.decodable_data_to_restore:
+                        del self.decodable_data_to_restore[key]
+                    super().__delitem__(key)
+
+            if not batched:
+                input_formatter.lazy_row_type = DictWithDecodableDataToRestore
+            else:
+                input_formatter.lazy_batch_type = DictWithDecodableDataToRestore
+
+        else:
+
+            class DictWithDecodableDataToRestore:
+                pass
```

Comment:
And this one

Reply:
I'm not sure I understand what's going on here tbh :/

Reply:
I think it's easiest to understand this with an example:
```python
dset = dset.with_format("torch") # torch-formatted dataset

def map_fn(ex):
     return {"shape": tuple(ex["image"].shape)}

dset = dset.map(map_fn)
```
The above example would save the "image" column as `torch.Tensor` (due to being accessed/decoded in `map_fn`) without this "restore" logic, which is not ideal because the image -> tensor -> image conversion is not always lossless. So what we do instead (using this special dict) is we store the "raw" (undecoded) values of decodable columns (only if they are decoded) to be able to restore them later (in the above case, we do this for the "image" column) unless these columns are overwritten with `__setitem__`  (e.g. `ex["image"] = "some val"`) in the transform.

Let me know if it's clear now.

Reply:
I see ! That's a super nice optimization trick :)

One nit:
what if the tensor is modified in place ? e.g.
```python
def map_fn(ex):
     ex["image"] += 1
     return ex
```
in that case the image data won't be updated.
IMO it's a case that it rare enough to not have to focus too much on it though - but documenting it would be appreciated.

Reply:
In this case, the correct behavior could be preserved by disabling lazy decoding for Numpy, TF, Jax and PyTorch. Moreover, the additional rationale is that if one sets the format to one of the above types before the map call, then it's OK to format all the `format_columns` eagerly (even if some of them are decodable) because we can assume these columns will be accessed in the map transform anyways. WDYT?

The options such as subclassing the tensor type or wrapping the tensor with an object that tracks method calls on it seem too error-prone.

Reply:
> The options such as subclassing the tensor type or wrapping the tensor with an object that tracks method calls on it seem too error-prone.

I agree - tbh I think we can ignore this for now and focus on this PR. As long as we document it it's fine